### PR TITLE
[NVPTX] Optimize v2x16 BUILD_VECTORs to PRMT

### PR DESCRIPTION
**Pull request by frasercrmck**
When two 16-bit values are combined into a v2x16 vector, and those values are truncated come from 32-bit values, a PRMT instruction can save registers by selecting bytes directly from the original 32-bit values. We do this during a post-legalize DAG combine, as these opportunities are typically only exposed after the BUILD_VECTOR's operands have been legalized.

Additionally, if the 32-bit values are right-shifted, we can fold in the shift by selecting higher bytes with PRMT. Only logical right-shifts by 16 are supported (for now) since those are the only situations seen in practice. Right shifts by 16 often come up during the legalization of EXTRACT_VECTOR_ELT.

This idea was brought up in a PR comment by @Artem-B.

### CODE REVIEW

File: llvm/test/CodeGen/NVPTX/bf16-instructions.ll [hunk 1]
```diff
@@ -159,8 +159,8 @@ define <2 x bfloat> @test_faddx2(<2 x bfloat> %a, <2 x bfloat> %b) #0 {
 ; SM70-LABEL: test_faddx2(
 ; SM70:       {
 ; SM70-NEXT:    .reg .pred %p<3>;
-; SM70-NEXT:    .reg .b16 %rs<13>;
-; SM70-NEXT:    .reg .b32 %r<24>;
+; SM70-NEXT:    .reg .b16 %rs<9>;
+; SM70-NEXT:    .reg .b32 %r<25>;
```
Comment (frasercrmck) [inline llvm/test/CodeGen/NVPTX/bf16-instructions.ll:163]: Any idea why this shows an increase in `%r` registers when it seems as though there are just as many as before?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 2]
```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
...
+      *Op = Op->getOperand(0);
+
+    if (Op->getValueType() != MVT::i16 || Op->getOpcode() != ISD::TRUNCATE ||
+        Op->getOperand(0).getValueType() != MVT::i32)
+      return SDValue();
+
+    *Op = Op->getOperand(0);
+
+    // Optionally, fold in a shift-right of the original operand and permute
+    // the two higher bytes from the shifted operand
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6211]: I'd reword it a bit ".. and let permute pick the two higher bytes of the original value directly".

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 3]
```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
...
+    if (Op->getOpcode() == ISD::SRL && isa<ConstantSDNode>(Op->getOperand(1))) {
+      if (cast<ConstantSDNode>(Op->getOperand(1))->getZExtValue() == 16) {
+        *OpBytes += 0x22;
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6214]: This could use a comment, that we're shifting PTMT byte selector to pick upper bytes instead of the lower ones.

I'd also be tempted to add an assert to make sure none of the permute selector values are out of range. E.g. in this case pre-incremented byte index would be expected to be 0x10 or 0x54 only.

**Review by Artem-B — APPROVED** [review]
Nice. LGTM with a few minor nits.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 4]
```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
...
+
+  // Check that each operand is an i16, truncated from an i32 operand. We'll
+  // select individual bytes from those original operands. Optionally, fold in a
+  // shift right of that original operand.
+  for (auto &[Op, OpBytes] : OpData) {
+    // Eat up any bitcast
+    if (Op->getOpcode() == ISD::BITCAST)
+      *Op = Op->getOperand(0);
+
+    if (Op->getValueType() != MVT::i16 || Op->getOpcode() != ISD::TRUNCATE ||
```
Comment (AlexMaclean) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6204]: It might be good to check that some of these instructions only have a single use, otherwise this transform is not a clear win and may increase register pressure.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 5]
```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
...
+    *Op = Op->getOperand(0);
+
+    // Optionally, fold in a shift-right of the original operand and permute
+    // the two higher bytes from the shifted operand
+    if (Op->getOpcode() == ISD::SRL && isa<ConstantSDNode>(Op->getOperand(1))) {
+      if (cast<ConstantSDNode>(Op->getOperand(1))->getZExtValue() == 16) {
+        *OpBytes += 0x22;
+        *Op = Op->getOperand(0);
+      }
+    }
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: This seems like a separate optimization to me: `prmt a (shr b [8|16|24]) x` -> `prmt a b y`.

Could we implement a separate DAG combine to do this? That way, `LowerBUILD_VECTOR` (and eventually `LowerINSERT_VECTOR_ELT` and `LowerEXTRACT_VECTOR_ELT` once they begin to use `prmt`) can use it.

File: llvm/test/CodeGen/NVPTX/i16x2-instructions.ll [hunk 6]
```diff
@@ -807,14 +807,11 @@ define <2 x i16> @test_select_cc_i16_i32(<2 x i16> %a, <2 x i16> %b,
...
-; COMMON-NEXT:    .reg .b16 %rs<3>;
-; COMMON-NEXT:    .reg .b32 %r<4>;
+; COMMON-NEXT:    .reg .b32 %r<5>;
 ; COMMON-EMPTY:
 ; COMMON-NEXT:  // %bb.0:
 ; COMMON-NEXT:    ld.param.v2.u32 {%r1, %r2}, [test_trunc_2xi32_param_0];
-; COMMON-NEXT:    cvt.u16.u32 %rs1, %r2;
-; COMMON-NEXT:    cvt.u16.u32 %rs2, %r1;
-; COMMON-NEXT:    mov.b32 %r3, {%rs2, %rs1};
+; COMMON-NEXT:    prmt.b32 %r3, %r1, %r2, 0x5410U;
```
Comment (justinfargnoli) [inline llvm/test/CodeGen/NVPTX/i16x2-instructions.ll:814]: What's the SASS diff for this change?

File: llvm/test/CodeGen/NVPTX/bf16-instructions.ll [hunk 7]
```diff
@@ -197,8 +196,7 @@ define <2 x bfloat> @test_faddx2(<2 x bfloat> %a, <2 x bfloat> %b) #0 {
 ; SM70-NEXT:    setp.nan.f32 %p2, %f6, %f6;
 ; SM70-NEXT:    or.b32 %r21, %r17, 4194304;
 ; SM70-NEXT:    selp.b32 %r22, %r21, %r20, %p2;
-; SM70-NEXT:    { .reg .b16 tmp; mov.b32 {tmp, %rs11}, %r22; }
-; SM70-NEXT:    mov.b32 %r23, {%rs11, %rs7};
+; SM70-NEXT:    prmt.b32 %r23, %r22, %r12, 0x7632U;
```
Comment (justinfargnoli) [inline llvm/test/CodeGen/NVPTX/bf16-instructions.ll:199]: What's the SASS diff for this change?

**Review by justinfargnoli — COMMENTED** [review]
Overall, this is great! Thanks for following-up on @Artem-B's comment :)

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 5 above)
Comment (justinfargnoli) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: Piggybacking off of this, if we do perform `prmt a (shr b [8|16|24]) x` -> `prmt a b y` separately, then should this be a custom lowering instead of a DAG combine?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 4 above)
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6204]: Nit: Also, it would be more readable to express the check as `if (!( Op->getValueType() == MVT::i16 && Op->getOpcode() == ISD::TRUNCATE && Op->getOperand(0).getValueType() == MVT::i32))` -- positive assertions are easier to reason about, compared to a bunch of  "not this" or "not that".

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 8]
```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
...
+  // Check that each operand is an i16, truncated from an i32 operand. We'll
+  // select individual bytes from those original operands. Optionally, fold in a
+  // shift right of that original operand.
+  for (auto &[Op, OpBytes] : OpData) {
+    // Eat up any bitcast
+    if (Op->getOpcode() == ISD::BITCAST)
+      *Op = Op->getOperand(0);
+
+    if (Op->getValueType() != MVT::i16 || Op->getOpcode() != ISD::TRUNCATE ||
+        Op->getOperand(0).getValueType() != MVT::i32)
```
[the remark this answers is not in the document]
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6205]: In order to use it in a `prmt` we would need to extend it to `i32`. That is not a clear win. With truncate we reach out to the original `i32` we can use in permute.

With that in mind, we may not need to check for the single use here -- we're not adding any new register uses, we're just short-circuiting a part of the graph. If there are multiple uses for both 16-bit input values, there may be a slight benefit if we emit combining `mov`, but in most of the cases, `prmt` would allow us to skip some i32->i16 conversions and would be a net gain. I do not think we're at risk of increasing register pressure if we ignore one-use-ness of the inputs as prmt will be using the already existing `i32` values.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 5 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: > This seems like a separate optimization to me: prmt a (shr b [8|16|24]) x -> prmt a b y.

One way we could generalize that is via two-step process:
- convert shifts and truncates by 8/16/24, vector extraction of i8/i16 into `prmt`, This give us convenient access to an i32 and a mask of valid bits in it. E.g. `shr b,16` -> `prmt b, 0, 0x4410`.
- then combine individual i32s with partial values into one. E.g. `or(prmt(a,0,0x4410), prmt(b,0,0x1044), 16))` -> `prmt(a,b,0x5410)`. It would be applicable to `prmt` itself, `or`, `build_vector`, `insert_vector_elt`, etc.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (justinfargnoli) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6205]: > In order to use it in a prmt we would need to extend it to i32. That is not a clear win. With truncate we reach out to the original i32 we can use in permute.

Ahh, I see. Thank you for the explanation.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 5 above)
Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: My original approach was custom-lower v2x16 BUILD_VECTORs to PRMT, but I found it was getting too complicated to identify the cases that were optimal to use with PRMT. As discussed in another comment thread, unless the original values are i32, this optimization broadly isn't worth it. During lowering of BUILD_VECTOR, we often don't as easily see the 16-bit-from-32-bit values that make it worthwhile. My concern with being too eager about BUILD_VECTOR, is that there will necessarily be cases where we need to undo a suboptimal PRMT, which feels wrong to me.

I imagine that Artem's idea could work, however. I suspect we'll still have to undo an eagerly matched PRMT, though. Perhaps we'd have to be more careful with one-use nodes?

File: llvm/test/CodeGen/NVPTX/i16x2-instructions.ll (hunk 6 above)
Comment (frasercrmck) [reply llvm/test/CodeGen/NVPTX/i16x2-instructions.ll:814]: On `ptxas v10.2.89` for SM70, no diff - it created a `PRMT` instruction either way.

On `ptxas v12.6.1`, for SM70,

``` diff
@@ -1,19 +1,11 @@
 test_trunc_2xi32:
  IMAD.MOV.U32 R1, RZ, RZ, c[0x0][0x28]
  @!PT SHFL.IDX PT, RZ, RZ, RZ, RZ
- LDC.U16 R0, c[0x0][0x160]
- MOV R3, c[0x0][0x164]
- LDC.U16 R5, c[0x0][0x164]
- PRMT R0, R0, 0x5410, R5
+ IMAD.MOV.U32 R3, RZ, RZ, c[0x0][0x164]
+ MOV R0, c[0x0][0x160]
+ PRMT R0, R0, 0x5410, R3
  LDC RZ, c[0x0][R3+0x160]
  EXIT
 .L_x_0:
  BRA `(.L_x_0)
- NOP
- NOP
- NOP
- NOP
- NOP
- NOP
- NOP
 .L_x_1:
```

I'm not sure if the fewer NOPs are spurious or not.

On `ptxas v12.6.1`, for SM89,

``` diff
@@ -1,23 +1,15 @@
 test_trunc_2xi32:
  IMAD.MOV.U32 R1, RZ, RZ, c[0x0][0x28]
- ULDC.U16 UR5, c[0x0][0x164]
- ULDC.U16 UR4, c[0x0][0x160]
- IMAD.U32 R3, RZ, RZ, UR5
- MOV R0, UR4
- ULDC UR4, c[0x0][0x164]
+ IMAD.MOV.U32 R3, RZ, RZ, c[0x0][0x164]
+ MOV R0, c[0x0][0x160]
  PRMT R0, R0, 0x5410, R3
- LDC RZ, cx[UR4][0x160]
+ LDC RZ, c[0x0][R3+0x160]
  EXIT
 .L_x_0:
  BRA `(.L_x_0)
  NOP
  NOP
  NOP
- NOP
- NOP
- NOP
- NOP
- NOP
  NOP
  NOP
  NOP
```

File: llvm/test/CodeGen/NVPTX/bf16-instructions.ll (hunk 7 above)
Comment (frasercrmck) [reply llvm/test/CodeGen/NVPTX/bf16-instructions.ll:199]: On `ptxas v12.6.1` for SM89 and SM70, no diff.
On `ptxas v10.2.89` for SM70,

``` diff
@@ -1,6 +1,6 @@
 test_faddx2:
  IMAD.MOV.U32 R1, RZ, RZ, c[0x0][0x28]
- NOP
+ @!PT SHFL.IDX PT, RZ, RZ, RZ, RZ
  LDC.U16 R0, c[0x0][0x166]
  IMAD.MOV.U32 R4, RZ, RZ, c[0x0][0x164]
  IMAD.MOV.U32 R7, RZ, RZ, c[0x0][0x160]
@@ -22,8 +22,14 @@
  IADD3 R3, R3, 0x7fff, R0
  @P0 LOP3.LUT R3, R0, 0x400000, RZ, 0xfc, !PT
  PRMT R2, R2, 0x7632, R3
- LDC RZ, c[0x0][R7]
+ LDC RZ, c[0x0][R7+0x160]
  EXIT
-.L_1:
- BRA `(.L_1)
-.L_20:
+.L_x_0:
+ BRA `(.L_x_0)
+ NOP
+ NOP
+ NOP
+ NOP
+ NOP
+ NOP
+.L_x_1:
```

I can't claim to know what's going on here and why `prmt` is causing it.

File: llvm/test/CodeGen/NVPTX/fma-relu-contract.ll [hunk 9]
```diff
@@ -1148,8 +1147,7 @@ define <2 x bfloat> @fma_bf16x2_expanded_no_nans_multiple_uses_of_fma(<2 x bfloa
 ; CHECK-SM70-NEXT:    setp.nan.f32 %p8, %f18, %f18;
 ; CHECK-SM70-NEXT:    or.b32 %r58, %r54, 4194304;
 ; CHECK-SM70-NEXT:    selp.b32 %r59, %r58, %r57, %p8;
-; CHECK-SM70-NEXT:    { .reg .b16 tmp; mov.b32 {tmp, %rs23}, %r59; }
-; CHECK-SM70-NEXT:    mov.b32 %r60, {%rs23, %rs20};
```
[the remark this answers is not in the document]
Comment (frasercrmck) [reply llvm/test/CodeGen/NVPTX/fma-relu-contract.ll:1152]: I don't think so because it's still used for `(trunc (srl s, 16))` or `(extractelt $vec, 1)`. Perhaps if we generally matched both of those to PRMTs we could remove the code, but I suspect we'll always need the option to fall back to these patterns.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6205]: Yep exactly. I was initially matching `i16` values but the generated PTX was always worse in that situation.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 4 above)
Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6204]: I'll write some tests to check what happens. Artem in another comment thinks that it might be okay if the instructions aren't one-use.

File: llvm/test/CodeGen/NVPTX/i16x2-instructions.ll (hunk 6 above)
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/i16x2-instructions.ll:814]: > I'm not sure if the fewer NOPs are spurious or not.

AFAICT, it's just padding to fill post-branch bubble and possibly align the `.L_x_1` address. NVIDIA does not disclose the microarchitecture details, so it's just a guess. In any case, those nops are not executed, so we do not care.

File: llvm/test/CodeGen/NVPTX/bf16-instructions.ll (hunk 7 above)
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/bf16-instructions.ll:199]: > @!PT SHFL.IDX PT, RZ, RZ, RZ, RZ

This appears to be a fancy `NOP` as the `@!PT` predicate will always be false.

> LDC RZ, c[0x0][R7+0x160]

The `LDC RZ` instruction appears to be part of some sort of calling convention magic that's not directly related to the calculations we do in the function -- it loads something from the constants/parameters area, and, apparently ignores the result. We can ignore that.

The short version -- the code didn't change.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 5 above)
Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: How do you see this idea working in the case of something like `generic_2xi16` [here](https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/NVPTX/load-store.ll#L197-L217)?

In this case, it's better to extract to actual 16-bit registers from the vector using `mov.b32 {%rs1, %rs2}, %r1`, because we're doing scalarized 16-bit operations on them. We can't detect that at the point at which we do `LowerEXTRACT_VECTOR_ELEMENT`. Well we can, but it's not common to have to look at uses when lowering nodes, and it's a bit of a red flag to me. The problem with eagerly creating PRMTs when lowering the vector extracts is that it gives us these awkward 32-bit registers, which we have to undo.

We'd probably have to do what `NVPTXDAGToDAGISel::tryEXTRACT_VECTOR_ELEMENT` is doing and try and find two `truncate`s from two `PRMT`s from the same original vector and replace them all with a single `NVPTX::I32ToV2I16`. This starts to sound a little strange, especially since we haven't proven that using PRMTs in this way is going to bring any tangible benefits.

What do you think?

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: The '32-bit-source + mask' intermadiate value approach would be able to preserve that nice example. Nothing stops us from lowering a pair of `SM(value, 0xffff), SM(value, 0xffff0000)` as `mov.s32 {rs1, rs2}, value` when we need two 16-bit values, but use `prmt` when the destination is 32-bit. E.g. if your example above would only swap those 16-bit parts, without addition.

Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: I suppose so, yes.

I was also wondering whether we'd need a way of encapsulating the idea of an "unused" byte for maximum effect, as in the example of truncation: `i16 = trunc (i32 a)` -> `prmt a, b, 0xXX10`. There's no concrete byte of `a` or `b` to select for the truncated-out bytes, nor any concrete value of `b` - both can be anything we want. If we choose any concrete values or indices, it might hinder or confuse optimal combines/lowering. Using something like `undef` as `b` would work, and maybe selecting any byte from an `undef` operand would give us enough info to know we can replace them with anything.

In any case, I won't have the time to dedicate to trying out any of these ideas, I'm afraid. If you want to take it on, go ahead! We could either merge this PR as-is, or just abandon it. I'm easy.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6217]: You can always use zero as one of the PRMT inputs, or sign-extend upper bits.

This particular change is in the nice-to-have category but it's not particularly important. It's more of a "how we could produce slightly better code with a cleaner/nicer implementation", but given that things already work reasonably well the upside is very limited. It may be a suitable starter project for someone getting up to speed with LLVM/NVPTX in the future.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 2 above)
Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6211]: Good idea, thanks

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 3 above)
Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6214]: Done, thanks

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 4 above)
Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6204]: I've tried some simple tests with multiple uses of the truncate and/or the original value.

When you reuse the truncate you appear to increase register pressure (though the SASS remains the same): https://godbolt.org/z/MT8PqshsW

When you reuse the original value the register pressure looks better, indicating the PRMT is worthwhile. Though the SASS is the same: https://godbolt.org/z/MqdbT6W59

When you reuse both, the register pressure is still worse, though the SASS remains the same: https://godbolt.org/z/Mc8W763M8

So, even though the SASS remains the same in these simple examples, it indicates we should probably bail out if the truncate has multiple uses. Multiple uses of the original value appears to be alright.

Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6204]: Oh yes I should add, those links are using ptxas v11.5.0. It's the same result with the most recent 12.6.1; the SASS is always the same between the PRMT and non-PRMT versions.

Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: I've pushed a change that forbids multiple uses of the truncate, and have added tests for it.

Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Done

**Comment by justinfargnoli** [comment]
> @[justinfargnoli](https://github.com/justinfargnoli) requested a review from [Artem-B](https://github.com/Artem-B) 

Apologies, I didn't mean to click the request review button.

**Review by justinfargnoli — APPROVED** [review]
The change itself LGTM :)

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 4 above)
Comment (justinfargnoli) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: > the register pressure is still worse, though the SASS remains the same

What register pressure are you referring to here? Is it the number of declared registers: 
```
.reg .b32       %r<number of registers declared>;
```

---

Do we know of a case where the SASS differs due to this change?

Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Yeah, I was referring to the number declared registers. I thought that's what you were referring to as it's the only thing we have "control" of.

I've only tried the examples I linked to above but no I've not seen a difference in the actual SASS. I could imagine that more PTX registers could manifest in worse SASS in more complex examples? It's not much to go on, though.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: On the tests we're consistently declaring four less `b16` registers, and one more `.b32`. It looks like a net win to me, even if it does not matter for SASS. It means we're doing less unnecessary stuff with `.b16` intermediate values.

**Review by Artem-B — APPROVED** [review]
LGTM.
